pyccel/parser/semantic.py

```diff
@@ -1380,6 +1385,83 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign, *
 
         return lhs
 
+    def _assign_GeneratorComprehension(self, expr, **settings):
+
+        result   = expr.expr
+        lhs_name = _get_name(expr.lhs)
+        lhs  = self.check_for_variable(lhs_name)
+
+        loop = expr.loops
+        loops = expr.loops
+        nlevels = 0
+        index   = Variable('int',self.get_new_name('to_delete'), is_temp=True)
+        self.insert_variable(index)
+        new_expr = []
+        while isinstance(loop, For):
+            nlevels+=1
+            iterable = Iterable(self._visit(loop.iterable, **settings))
+            n_index = max(1, iterable.num_indices_required)
+            iterable.set_indices(*[index]*n_index)
```

Comment:
Actually this is unrelated to the previous comment. Here this is not supposed to work. The generated expression won't make any sense. It is simply a trick to avoid creating lots of indexes which need to be deleted later. This is done to handle expressions such as:
```
sum(aij for ai in a for aij in ai)
```
In this case the central expression is `ai`, in order to know the dtype of this object we need it to have been defined. To define it we must be able to call `iterable.get_target_from_range` and deduce the dtype of an example element. In this case the element that is returned (the variable `iterator_rhs`) is the element `a[to_delete,to_delete]`. The fact that the same index is used multiple times is not a problem as the target that is actually used is only created when we visit the for loop. This stage only serves 1 purpose : to determine the dtype of the final result in order to be able to initialise the result variable

Reply:
I see. Could you please add a comment for developers?

Reply:
Done 201a701